src/Libraries/Microsoft.Extensions.Telemetry.Testing/Metering/MetricCollector.cs

```diff
@@ -2,446 +2,540 @@
-    /// </returns>
-    public MetricValuesHolder<T>? GetObservableCounterValues<T>(string observableCounterName)
-        where T : struct => GetInstrumentCapturedData<T>(_allObservableCounters, observableCounterName);
+    /// <param name="clear">Setting this to <see langword="true"/> will atomically clear the set of accumulated measurements.</param>
+    /// <returns>The metric values recorded by this collector.</returns>
+    public IReadOnlyCollection<MetricValue<T>> SnapshotMetricValues(bool clear = false)
+    {
+        lock (_measurements)
+        {
+            ThrowIfDisposed();
+
+            // copy the values
+            var values = new MetricValue<T>[_metrics.Count];
+            int index = 0;
+            foreach (var metric in _metrics.Values)
+            {
+                values[index++] = new MetricValue<T>(metric.Value, metric.Tags, metric.NumWrites);
+            }
+
+            if (clear)
+            {
+                InnerClear();
+            }
+
+            return values;
+        }
+    }
 
     /// <summary>
-    /// Gets the object containing all the captured metering data that has been recorded by an observable updown counter instrument.
+    /// Gets the latest measurement collected, if any.
     /// </summary>
-    /// <typeparam name="T">The type of metric measurement value.</typeparam>
-    /// <param name="observableUpDownCounterName">The metric name.</param>
-    /// <returns>
-    /// The <see cref="MetricValuesHolder{T}"/> instance or <see langword="null"/> if an observable updown counter with given <paramref name="observableUpDownCounterName"/> was not found.
-    /// </returns>
-    public MetricValuesHolder<T>? GetObservableUpDownCounterValues<T>(string observableUpDownCounterName)
-        where T : struct => GetInstrumentCapturedData<T>(_allObservableUpDownCounters, observableUpDownCounterName);
+    public CollectedMeasurement<T>? LatestMeasurementTaken
+    {
+        get
+        {
+            lock (_measurements)
+            {
+                ThrowIfDisposed();
+                return _measurements.Count > 0 ? _measurements[_measurements.Count - 1] : null;
+            }
+        }
+    }
 
     /// <summary>
-    /// Gets a measurement value recorded by a counter instrument.
+    /// Gets the latest metric value that was updated.
     /// </summary>
-    /// <typeparam name="T">The type of metric measurement value.</typeparam>
-    /// <param name="counterName">The metric name.</param>
-    /// <param name="tags">The dimensions collection describing the measurement value.</param>
-    /// <returns>The measurement value or <see langword="null"/> if the value was not recorded.</returns>
-    public T? GetCounterValue<T>(string counterName, params KeyValuePair<string, object?>[] tags)
-        where T : struct => GetCounterValues<T>(counterName)?.GetValue(tags);
+    public MetricValue<T>? LatestMetricValueUpdated
```

Comment:
How important do you feel it is to have the collector producing both measurements and metrics? 

The API defines the metric value of counters as the sum of measurements and for everything else it is identical to the last measurement value. Claiming the last measurement value of a histogram is the metric value feels like a very strange definition that is picked solely to force it to use the same MetricValue\<T> type as other instrument types.

Could we eliminate the MetricValue concept on this API and instead work with lists of measurements, last measurements, and if needed sums of measurements? I believe for all the tests modified by this PR other than tests of MetricCollector itself, they would all pass if LatestMetricValueUpdated was substituted with LatestMeasurementTaken. There were a couple tests that tested LatestMetricValueUpdated.Value on a Counter instrument but the expected value was always 1. I assume it would be most common when testing a counter to do some action in the unit test that increments the counter once in which case the sum is also equal to the last measurement.

Reply:
All right, I nuked the MetricValue type and associated logic, and introduced instead a few extension methods to simplify manipulating lists of meaaurements.